openhtf/io/station_api.py

```diff
@@ -0,0 +1,574 @@
+                                         timeout_s=timeout_s,
+                                         **MULTICAST_KWARGS()):
+      try:
+        yield cls(StationInfo(host, **json.loads(response)))
+      except ValueError:
+        _LOG.debug('Received malformed JSON from %s: %s', host, response)
+      except TypeError:
+        _LOG.debug('Received invalid discovery response from %s: %s',
+                   host, response, exc_info=True)
+
+  @property
+  def tests(self):
+    """List active Test instances on this station."""
+    print('listing tests')
+    with self._shared_proxy._lock:
+      print('acquired lock')
+      tests = self._shared_proxy.list_tests()
+    print('received %s tests' % len(tests))
+
+    updated_tests = {}
+    for test_dict in tests:
+      test_uid = test_dict['test_uid']
+      ctime_millis = test_dict['created_time_millis']
+      if (test_uid not in self._known_tests or
+          ctime_millis != self._known_tests[test_uid].created_time_millis):
+        updated_tests[test_uid] = RemoteTest(
+            self.make_proxy, self._shared_proxy, self._history, **test_dict)
+      else:
+        updated_tests[test_uid] = self._known_tests[test_uid]
+
+    self._known_tests = updated_tests
+    return self._known_tests.itervalues()
+
+
+### Server-side objects below here. ###
+
+
+class StationApi(object):
+
+  UID = '%s:%s' % (os.getpid(), util.TimeMillis())
+
+  def list_tests(self):
+    """List currently known test types.
+
+    A new 'test type' is created each time openhtf.Test is instantiated, and
+    lasts as long as there are any external references to it (ie, outside the
+    internal tracking structures within OpenHTF).
+
+    This means creating large numbers of openhtf.Test instances and keeping
+    references to them around can cause memory usage to grow rapidly and
+    station_api performance to degrade; don't do that.
+
+    Returns:
+      List of RemoteTest tuple values (as a dict).
+    """
+    retval = [{
+        'test_uid': test.uid,
+        'test_name': test.GetOption('name'),
+        'created_time_millis': long(test.created_time_millis),
+        'last_run_time_millis':
+            test.last_run_time_millis and long(test.last_run_time_millis),
+    } for test in openhtf.Test.TEST_INSTANCES.values()]
+    _LOG.debug('RPC:list_tests() -> %s results', len(retval))
+    return retval
+
+  @staticmethod
+  def _serialize_state_dict(state, start_time_millis=None,
+                            skip_phases=0, skip_logs=0):
+    if (start_time_millis and
+        start_time_millis == state['test_record'].start_time_millis):
+      state['test_record'] = mutablerecords.CopyRecord(state['test_record'])
+      del state['test_record'].phases[:skip_phases]
+      del state['test_record'].log_records[:skip_logs]
+
+    return {
+        'status': state['status'].name,
+        'test_record': saxutils.escape(pickle.dumps(state['test_record'])),
```

Comment:
Might prefer wrapping this as `Binary` rather than escaping it: https://docs.python.org/2/library/xmlrpclib.html#xmlrpclib.Binary

That way `xmlrpclib` will escape it only when necessary